### ISSUE TITLE
Loss/accuracy evaluation during training (training=True) and after training (training=False) differs significantly for training set

### ISSUE THREAD

**Issue by neudorf** [danielegrattarola/spektral#417]
Hi again, thank you for your help with a previous issue. I have noticed some strange behavior using GeneralGNN, where the loss/accuracy for the training set reported during training differs widely from the evaluated loss/accuracy after training. You can see this behavior in the example you have posted for graph prediction, `general_gnn.py`, by running the example completely as is, and then adding the following lines to evaluate the performance on the training set:
```python
loader_tr = DisjointLoader(data_tr, batch_size=batch_size)
results_tr = evaluate(loader_tr)
print("Final train results - Loss: {:.3f} - Acc: {:.3f}".format(*results_tr))
```
Training loss/accuracy reported for the final epoch (as calculated by the `train_step` function) is `Ep. 50 - Loss: 0.257 - Acc: 0.895`, but when using the `evaluate` function on the training data as shown above, I get `Final train results - Loss: 0.912 - Acc: 0.479`.
I believe the issue centers around the use of `training=False` for `evaluate`, but I can't use `training=True` or it will keep training the model when I just want to evaluate it. I am getting the same issue as well when I use the Keras built in Model training APIs (model.fit, model.evaluate).
I suspect the issue may be related to how Spektral interacts with `BatchNormalization`, as if I turn batch normalization off (`model = GeneralGNN(data.n_labels, activation="softmax", batch_norm=False)`) I get this for the final epoch loss/accuracy on the training data `Ep. 50 - Loss: 0.527 - Acc: 0.756` and this by running `evaluate` on the training data `Final train results - Loss: 0.501 - Acc: 0.770`. This is much more comparable, but obviously not ideal as we take a big hit to accuracy by turning off batch normalization.
Thank you for your insight, much appreciated.

**Comment by danielegrattarola** [comment]
Hey, I'm not sure what the issue is here so I will investigate. 
However, `training=True` does not mean that you are training the model (as in changing the parameters), but only that you are using the model in "training mode", ie, with dropout and by using the batch statistics for batchnorm. 
If you set `training=False` then Keras turns off dropout and uses the learned statistics for batchnorm.

**Comment by neudorf** [comment]
Great, thank you for looking into it.
OK, I was doing some testing and thought I saw the weights changing when using `training=True`, thanks for clarifying this.

**Comment by neudorf** [comment]
Hey Daniele, I just wanted to touch base to check if you can see a way forward for fixing this issue?

**Comment by danielegrattarola** [comment]
Hey, 

I've looked at it but this is most likely a Keras issue, not Spektral. Without seeing the rest of the code it's hard to make a call.
Sorry about this.

**Comment by neudorf** [comment]
OK, thank you for looking into it. This problem occurs in the `general_gnn.py` example that you have posted, how would you approach fixing this problem for the example as it is? Do you have ideas/plans to implement your own batch normalization for Spektral, or is Spektral not going to support batch normalization?

**Comment by danielegrattarola** [comment]
This is not a Spektral issue so it's not really a matter of supporting anything. 
Here's a possible explanation for what's happening: https://stackoverflow.com/questions/50047653/set-training-false-of-tf-layers-batch-normalization-when-training-will-get-a

I guess that BN is simply not a good choice for that particular dataset in the example, so it might be better to simply disable it in general_gnn.py.

**Comment by neudorf** [comment]
Thanks for the link, I'll see if that helps with my data.
